src/main/java/com/hivemq/mqtt/message/pool/FreeIdRanges.java

```diff
@@ -0,0 +1,198 @@
+ * When the ID is returned, it either joins one of the existing {@link Range} intervals in the list (if it is adjacent)
+ * or it forms a new {@link Range} that is added to the list.
+ * <p>
+ * This class is NOT thread-safe.
+ * <a
+ * href="https://github.com/hivemq/hivemq-mqtt-client/blob/master/src/main/java/com/hivemq/client/internal/util/Ranges.java">The
+ * original implementation in the HiveMQ Java Client.</a>
+ */
+public class FreeIdRanges {
+
+    private static final @NotNull Logger log = LoggerFactory.getLogger(FreeIdRanges.class);
+
+    final int minAllowedId;
+    final int maxAllowedId;
+
+    private @NotNull Range rootRange;
+
+    public FreeIdRanges(final int minId, final int maxId) {
+        minAllowedId = minId;
+        maxAllowedId = maxId;
+        rootRange = new Range(minId, maxId + 1);
+    }
+
+    /**
+     * Provides a new ID that is not currently allocated.
+     *
+     * @return a new ID if available in any of the ranges or {@link NoMessageIdAvailableException} if ran out of IDs.
+     */
+    public int takeNextId() throws NoMessageIdAvailableException {
+        if (rootRange.start == rootRange.end) {
+            throw new NoMessageIdAvailableException();
+        }
+
+        final int id = rootRange.start;
+        rootRange.start++;
+        if ((rootRange.start == rootRange.end) && (rootRange.next != null)) {
+            rootRange = rootRange.next;
+        }
+        return id;
+    }
+
+    /**
+     * Provides the requested ID if it is available or some other ID otherwise.
+     *
+     * @param id an ID that the caller attempts to take.
+     * @return the requested {@param id} if it is available in one of the {@link Range} intervals or otherwise some
+     *         other free ID.
+     */
+    public int takeIfAvailable(final int id) throws NoMessageIdAvailableException {
+        if (id < minAllowedId || id > maxAllowedId) {
+            log.warn("Attempting to take an ID {} that is outside the valid range [{}, {}], will try taking another ID.",
+                    id,
+                    minAllowedId,
+                    maxAllowedId);
+        }
+
+        Range current = rootRange;
+        Range prev = null;
+
+        while (current != null) {
+            if (id >= current.start && id < current.end) {
+
+                final int prevCurStart = current.start;
+                current.start = id + 1;
+                final Range lowerRange = prevCurStart == id ? null : new Range(prevCurStart, id, current);
+
+                if (lowerRange != null) {
+                    if (prev != null) {
+                        prev.next = lowerRange;
+                    } else {
+                        rootRange = lowerRange;
+                    }
+                }
+
+                return id;
+            }
+
+            prev = current;
+            current = current.next;
+        }
+
+        return takeNextId();
```

Comment:
The result of takeIfAvailable() is not accepted when it is a different one then the requested one. This method does not need to return a packet id in such case.

Reply:
yes, the handling of taking/returning the id is weird there. refactored a bit